### PR TITLE
add accesslog to enhance obversiblity of kmesh

### LINKED ISSUE #643: add accesslog for Kmesh L4
**What would you like to be added**:
Use ringbuf to upload metrics needed by accesslog for aggregated publishing in user space
**Why is this needed**:
Improving observability for Kmesh L4

### PR DESCRIPTION
**Pull request by LiZhenCheng9527**
**What type of PR is this?**
/kind enhancement
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #643 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
can you paste what accesslog look like with this patch?

**Comment by LiZhenCheng9527** [comment]
time="2024-08-14T08:12:20Z" level=info msg="2024-08-14 08:12:21.257461842 +0000 UTC src.addr=10.244.0.7:46268, src.workload=sleep-bc9998558-x2t89, src.namespace=default, dst.addr=10.244.0.8:80, dst.service=httpbin.default.svc.cluster.local, dst.workload=httpbin-5c5944c58c-7bvzk, dst.namespace=default, direction=INBOUND, sent_bytes=3, received_bytes=87, duration=1.167347ms" subsys=pkg/telemetry

File: pkg/controller/telemetry/metric.go [hunk 1]
```diff
@@ -197,19 +203,25 @@ func (m *MetricController) Run(ctx context.Context, mapOfTcpInfo *ebpf.Map) {
...
 				serviceLabels.reporter = "destination"
+				accesslog.direction = "INBOUND"
 			}
 			if data.direction == constants.OUTBOUND {
 				workloadLabels.reporter = "source"
 				serviceLabels.reporter = "source"
+				accesslog.direction = "OUTBOUND"
 			}
 
+			if data.state == TCP_CLOST {
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/metric.go:222]: TCP_CLOSED?

File: pkg/controller/telemetry/accesslog_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			got := buildAccesslog(tt.args.data, tt.args.accesslog)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func Test_getOSBootTime(t *testing.T) {
+	t.Run("function test", func(t *testing.T) {
+		data := getOSBootTime()
+		fmt.Printf("%v", data)
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/accesslog_test.go:77]: I don't quite understand what this test means, maybe we should check the format?

File: pkg/controller/telemetry/accesslog.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	sourceInfo := fmt.Sprintf("src.addr=%s, src.workload=%s, src.namespace=%s", accesslog.sourceAddress, accesslog.sourceWorkload, accesslog.sourceNamespace)
+	destinationInfo := fmt.Sprintf("dst.addr=%s, dst.service=%s, dst.workload=%s, dst.namespace=%s", accesslog.destinationAddress, accesslog.destinationService, accesslog.destinationWorkload, accesslog.destinationNamespace)
+	connectionInfo := fmt.Sprintf("direction=%s, sent_bytes=%d, received_bytes=%d, duration=%vms", accesslog.direction, data.sentBytes, data.receivedBytes, (float64(data.duration) / 1000000.0))
+
+	logResult := fmt.Sprintf("%s %s, %s, %s", timeInfo, sourceInfo, destinationInfo, connectionInfo)
+	log.Infof("%s", logResult)
+	return logResult
+}
+
+func getOSBootTime() time.Time {
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/accesslog.go:58]: Why not return an error if this function fails?

File: pkg/controller/telemetry/accesslog.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	destinationNamespace string
+}
+
+func RunAccesslog(data requestMetric, accesslog logInfo) {
+	_ = buildAccesslog(data, accesslog)
+}
+
+func buildAccesslog(data requestMetric, accesslog logInfo) string {
+	closeTime := data.closeTime
+	startTime := getOSBootTime()
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/accesslog.go:45]: OS Boot time remains unchanged during program execution, so there is no need to execute it every time when print the access log.

File: pkg/controller/telemetry/accesslog.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	uptime := calculateUptime(startTime, closeTime)
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/accesslog.go:46]: What confuses me is does uptime refer to the current time? Why is it calculated like this?

File: pkg/controller/telemetry/accesslog.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+	timeInfo := fmt.Sprintf("%v", uptime)
+	sourceInfo := fmt.Sprintf("src.addr=%s, src.workload=%s, src.namespace=%s", accesslog.sourceAddress, accesslog.sourceWorkload, accesslog.sourceNamespace)
+	destinationInfo := fmt.Sprintf("dst.addr=%s, dst.service=%s, dst.workload=%s, dst.namespace=%s", accesslog.destinationAddress, accesslog.destinationService, accesslog.destinationWorkload, accesslog.destinationNamespace)
+	connectionInfo := fmt.Sprintf("direction=%s, sent_bytes=%d, received_bytes=%d, duration=%vms", accesslog.direction, data.sentBytes, data.receivedBytes, (float64(data.duration) / 1000000.0))
+
+	logResult := fmt.Sprintf("%s %s, %s, %s", timeInfo, sourceInfo, destinationInfo, connectionInfo)
+	log.Infof("%s", logResult)
+	return logResult
```
Comment (YaoZengzeng) [inline pkg/controller/telemetry/accesslog.go:55]: We have already printed the log, is it necessary to return `logResult`?

If it is for testing, maybe you can return `logResult` and print it externally.

File: pkg/controller/telemetry/accesslog_test.go (hunk 2 above)
Comment (LiZhenCheng9527) [reply pkg/controller/telemetry/accesslog_test.go:77]: This test is just to make sure that the getOs functions all execute correctly. 
Maybe I should update the test after return the error.

File: pkg/controller/telemetry/accesslog.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	sourceInfo := fmt.Sprintf("src.addr=%s, src.workload=%s, src.namespace=%s", accesslog.sourceAddress, accesslog.sourceWorkload, accesslog.sourceNamespace)
+	destinationInfo := fmt.Sprintf("dst.addr=%s, dst.service=%s, dst.workload=%s, dst.namespace=%s", accesslog.destinationAddress, accesslog.destinationService, accesslog.destinationWorkload, accesslog.destinationNamespace)
+	connectionInfo := fmt.Sprintf("direction=%s, sent_bytes=%d, received_bytes=%d, duration=%vms", accesslog.direction, data.sentBytes, data.receivedBytes, (float64(data.duration) / 1000000.0))
+
+	logResult := fmt.Sprintf("%s %s, %s, %s", timeInfo, sourceInfo, destinationInfo, connectionInfo)
+	return logResult
+}
+
+func getOSBootTime() (time.Time, error) {
+	cmd := exec.Command("uptime", "-s")
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/controller/telemetry/accesslog.go:58]: like syscall.SysInfo

**Pull request merged**